File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/severity/upstream_context_boundary.py
from __future__ import annotations

GROUP_PATHWAYS = {
    "POSITIVE_DISEASE_NEGATIVE_THERAPEUTIC_GROUP",
    "NEGATIVE_DISEASE_POSITIVE_THERAPEUTIC_GROUP",
}

NON_POTENCY_PATHWAYS = {"NEUTRAL_NON_POTENCY", "SUPPORT_ONLY_NON_POTENCY"}

UNRESOLVED_POLARITY_PATHWAYS = {
    "UNRESOLVED_NO_CASCADE",
    "POLARITY_CONTRADICTORY",
    "BLOCKED_BY_SAFETY_GATE",
    "NOT_EVALUATED",
}

RESOLVED_PHASE_STATUSES = {
    "RESOLVED_BY_DAY_BAND",
    "RESOLVED_BY_EVIDENCE",
    "RESOLVED_BY_CROSS_BOUNDARY_OVERRIDE",
}

UNRESOLVED_PHASE_STATUSES = {
    "PHASE_AMBIGUOUS",
    "PHASE_CONTRADICTORY",
    "PHASE_TARGET_CONTRADICTORY",
    "MISSING_EVIDENCE",
    "NOT_EVALUATED",
}


def _phase_is_resolved(phase_slot: dict | None) -> bool:
    if not phase_slot:
        return False
    return phase_slot.get("phase_status") in RESOLVED_PHASE_STATUSES


def _phase_is_unresolved(phase_slot: dict | None) -> bool:
    if not phase_slot:
        return False
    return phase_slot.get("phase_status") in UNRESOLVED_PHASE_STATUSES


def _derive_upstream_status(
    pathway: str | None,
    phase_slot: dict | None,
    polarity_provided: bool,
    phase_provided: bool,
) -> str:
    if not polarity_provided and not phase_provided:
        return "NOT_EVALUATED"

    if pathway and pathway in UNRESOLVED_POLARITY_PATHWAYS:
        return "AUDIT_ONLY_UPSTREAM_UNRESOLVED"

    if pathway and pathway in NON_POTENCY_PATHWAYS:
        return "AUDIT_ONLY_NON_POTENCY_CONTEXT"

    if pathway and pathway in GROUP_PATHWAYS:
        if phase_provided and _phase_is_unresolved(phase_slot):
            return "AUDIT_ONLY_UPSTREAM_UNRESOLVED"
        if phase_provided and _phase_is_resolved(phase_slot):
            return "READY_FOR_FUTURE_GATE_EVALUATION"
        if not phase_provided:
            return "AUDIT_ONLY_UPSTREAM_UNRESOLVED"

    if phase_provided and _phase_is_unresolved(phase_slot):
        return "AUDIT_ONLY_UPSTREAM_UNRESOLVED"

    return "NOT_EVALUATED"


def _find_polarity_slot(polarity: dict | None, slot: dict) -> dict | None:
    if not polarity:
        return None
    for r in polarity.get("slot_routings") or []:
        if r.get("formula_slot_id") == slot["formula_slot_id"] and (
            r.get("formula_target_id") is None
            or r.get("formula_target_id") == slot.get("formula_target_id")
        ):
            return r
    return None


def _find_phase_slot(phase: dict | None, slot: dict) -> dict | None:
    if not phase:
        return None
    for r in phase.get("slot_resolutions") or []:
        if r.get("formula_slot_id") == slot["formula_slot_id"] and (
            r.get("formula_target_id") is None
            or r.get("formula_target_id") == slot.get("formula_target_id")
        ):
            return r
    return None


def apply_upstream_context_boundary(
    slots: list[dict],
    polarity_routing: dict | None,
    phase_resolution: dict | None,
) -> list[dict]:
    polarity_provided = polarity_routing is not None
    phase_provided = phase_resolution is not None

    if not polarity_provided and not phase_provided:
        return [{**s, "upstream_context_status": "NOT_EVALUATED"} for s in slots]

    out: list[dict] = []
    for slot in slots:
        pol = _find_polarity_slot(polarity_routing, slot)
        ph = _find_phase_slot(phase_resolution, slot)

        if polarity_provided and pol is None:
            out.append(
                {
                    **slot,
                    "upstream_context_status": "NOT_EVALUATED",
                    "severity_status": "NOT_EVALUATED",
                    "severity_score": None,
                    "severity_band": None,
                    "severity_resolution_source": "NONE",
                    "selected_cascade": None,
                    "selected_dilution": None,
                    "reason_codes": sorted(
                        set([*slot["reason_codes"], "UPSTREAM_TARGET_POLARITY_NOT_RESOLVED"])
                    ),
                    "limitation_codes": sorted(
                        set([*slot["limitation_codes"], "PHASE6_UPSTREAM_CONTEXT_BOUNDARY"])
                    ),
                }
            )
            continue

        upstream_context_status = _derive_upstream_status(
            pol.get("pathway") if pol else None,
            ph,
            polarity_provided,
            phase_provided,
        )

        limitation_codes = list(slot["limitation_codes"])
        if upstream_context_status in {
            "AUDIT_ONLY_NON_POTENCY_CONTEXT",
            "AUDIT_ONLY_UPSTREAM_UNRESOLVED",
        }:
            limitation_codes = sorted(
                set([*limitation_codes, "PHASE6_UPSTREAM_CONTEXT_BOUNDARY"])
            )

        out.append(
            {
                **slot,
                "upstream_context_status": upstream_context_status,
                "selected_cascade": None,
                "selected_dilution": None,
                "limitation_codes": sorted(set(limitation_codes)),
            }
        )
    return out

File: packages/clinical-contracts/src/rule4/severity/upstreamContextBoundary.ts
import type { Rule4PhaseAdapterOutput } from '../phase/types.js';
import type { Rule4PolarityAdapterOutput } from '../polarity/types.js';
import type { Rule4SlotSeverityResolution } from './types.js';

export const RULE4_UPSTREAM_CONTEXT_STATUS_VALUES = [
  'READY_FOR_FUTURE_GATE_EVALUATION',
  'AUDIT_ONLY_NON_POTENCY_CONTEXT',
  'AUDIT_ONLY_UPSTREAM_UNRESOLVED',
  'NOT_EVALUATED',
] as const;

export type Rule4UpstreamContextStatus = (typeof RULE4_UPSTREAM_CONTEXT_STATUS_VALUES)[number];

const GROUP_PATHWAYS = new Set([
  'POSITIVE_DISEASE_NEGATIVE_THERAPEUTIC_GROUP',
  'NEGATIVE_DISEASE_POSITIVE_THERAPEUTIC_GROUP',
]);

const NON_POTENCY_PATHWAYS = new Set(['NEUTRAL_NON_POTENCY', 'SUPPORT_ONLY_NON_POTENCY']);

const UNRESOLVED_POLARITY_PATHWAYS = new Set([
  'UNRESOLVED_NO_CASCADE',
  'POLARITY_CONTRADICTORY',
  'BLOCKED_BY_SAFETY_GATE',
  'NOT_EVALUATED',
]);

function phaseIsResolved(
  phase: Rule4PhaseAdapterOutput['slotResolutions'][0] | undefined,
): boolean {
  if (!phase) return false;
  return (
    phase.phaseStatus === 'RESOLVED_BY_DAY_BAND' ||
    phase.phaseStatus === 'RESOLVED_BY_EVIDENCE' ||
    phase.phaseStatus === 'RESOLVED_BY_CROSS_BOUNDARY_OVERRIDE'
  );
}

function phaseIsUnresolved(
  phase: Rule4PhaseAdapterOutput['slotResolutions'][0] | undefined,
): boolean {
  if (!phase) return false;
  return (
    phase.phaseStatus === 'PHASE_AMBIGUOUS' ||
    phase.phaseStatus === 'PHASE_CONTRADICTORY' ||
    phase.phaseStatus === 'PHASE_TARGET_CONTRADICTORY' ||
    phase.phaseStatus === 'MISSING_EVIDENCE' ||
    phase.phaseStatus === 'NOT_EVALUATED'
  );
}

function deriveUpstreamStatus(
  pathway: string | undefined,
  phaseSlot: Rule4PhaseAdapterOutput['slotResolutions'][0] | undefined,
  polarityProvided: boolean,
  phaseProvided: boolean,
): Rule4UpstreamContextStatus {
  if (!polarityProvided && !phaseProvided) {
    return 'NOT_EVALUATED';
  }

  if (pathway && UNRESOLVED_POLARITY_PATHWAYS.has(pathway)) {
    return 'AUDIT_ONLY_UPSTREAM_UNRESOLVED';
  }

  if (pathway && NON_POTENCY_PATHWAYS.has(pathway)) {
    return 'AUDIT_ONLY_NON_POTENCY_CONTEXT';
  }

  if (pathway && GROUP_PATHWAYS.has(pathway)) {
    if (phaseProvided && phaseIsUnresolved(phaseSlot)) {
      return 'AUDIT_ONLY_UPSTREAM_UNRESOLVED';
    }
    if (phaseProvided && phaseIsResolved(phaseSlot)) {
      return 'READY_FOR_FUTURE_GATE_EVALUATION';
    }
    if (!phaseProvided) {
      return 'AUDIT_ONLY_UPSTREAM_UNRESOLVED';
    }
  }

  if (phaseProvided && phaseIsUnresolved(phaseSlot)) {
    return 'AUDIT_ONLY_UPSTREAM_UNRESOLVED';
  }

  return 'NOT_EVALUATED';
}

function findPolaritySlot(
  polarity: Rule4PolarityAdapterOutput | null | undefined,
  slot: Rule4SlotSeverityResolution,
) {
  return polarity?.slotRoutings.find(
    (r) =>
      r.formulaSlotId === slot.formulaSlotId &&
      (r.formulaTargetId == null || r.formulaTargetId === slot.formulaTargetId),
  );
}

function findPhaseSlot(
  phase: Rule4PhaseAdapterOutput | null | undefined,
  slot: Rule4SlotSeverityResolution,
) {
  return phase?.slotResolutions.find(
    (r) =>
      r.formulaSlotId === slot.formulaSlotId &&
      (r.formulaTargetId == null || r.formulaTargetId === slot.formulaTargetId),
  );
}

/**
 * Phase 6 upstream polarity/phase context boundary — audit-only; not potency eligibility.
 */
export function applyUpstreamContextBoundary(
  slots: Rule4SlotSeverityResolution[],
  polarityRouting: Rule4PolarityAdapterOutput | null | undefined,
  phaseResolution: Rule4PhaseAdapterOutput | null | undefined,
): Rule4SlotSeverityResolution[] {
  const polarityProvided = polarityRouting != null;
  const phaseProvided = phaseResolution != null;

  if (!polarityProvided && !phaseProvided) {
    return slots.map((s) => ({
      ...s,
      upstreamContextStatus: 'NOT_EVALUATED' as const,
    }));
  }

  return slots.map((slot) => {
    const pol = findPolaritySlot(polarityRouting, slot);
    const ph = findPhaseSlot(phaseResolution, slot);

    if (polarityProvided && !pol) {
      return {
        ...slot,
        upstreamContextStatus: 'NOT_EVALUATED' as const,
        severityStatus: 'NOT_EVALUATED' as const,
        severityScore: null,
        severityBand: null,
        severityResolutionSource: 'NONE' as const,
        selectedCascade: null,
        selectedDilution: null,
        reasonCodes: [
          ...new Set([...slot.reasonCodes, 'UPSTREAM_TARGET_POLARITY_NOT_RESOLVED']),
        ].sort(),
        limitationCodes: [
          ...new Set([...slot.limitationCodes, 'PHASE6_UPSTREAM_CONTEXT_BOUNDARY']),
        ].sort(),
      };
    }

    const upstreamContextStatus = deriveUpstreamStatus(
      pol?.pathway,
      ph,
      polarityProvided,
      phaseProvided,
    );

    let limitationCodes = [...slot.limitationCodes];
    if (
      upstreamContextStatus === 'AUDIT_ONLY_NON_POTENCY_CONTEXT' ||
      upstreamContextStatus === 'AUDIT_ONLY_UPSTREAM_UNRESOLVED'
    ) {
      limitationCodes = [...new Set([...limitationCodes, 'PHASE6_UPSTREAM_CONTEXT_BOUNDARY'])];
    }

    return {
      ...slot,
      upstreamContextStatus,
      selectedCascade: null,
      selectedDilution: null,
      limitationCodes: [...new Set(limitationCodes)].sort(),
    };
  });
}
